Make IOManager use Module instead of Method.

shoumikhin opened the pull request:
Summary: Let's not expose Method from Module so that it's not getting misused beyond its owner.

Differential Revision: D80595261

mergennachin commented on extension/llm/runner/io_manager/io_manager.h:
```diff
@@ -144,13 +176,21 @@ class ET_EXPERIMENTAL IOManager {
+  }
+
+private:
+  ET_MODULE_NAMESPACE::Module& module_;
```
Can it be `const ET_MODULE_NAMESPACE::Module& module_;`?

mergennachin commented on extension/llm/runner/io_manager/io_manager.h:
```diff
@@ -29,6 +25,10 @@ namespace llm {
 class ET_EXPERIMENTAL IOManager {
  public:
+
+  explicit IOManager(ET_MODULE_NAMESPACE::Module& module)
+  : module_(module) {}
```
IOManager holds a reference but doesn't ensure Module outlives it.

Perhaps use shared_ptr<Module> or weak_ptr<Module> 
```
  explicit IOManager(std::shared_ptr<Module> module) : module_(module) {}
  // Or

  explicit IOManager(std::weak_ptr<Module> module) : module_(module) {
    ET_CHECK_MSG(!module.expired(), "Module must be valid");
  }
```

mergennachin replied:
Also should you test if module is loaded in the constructor?

mergennachin commented on extension/llm/runner/io_manager/io_manager.h:
```diff
@@ -38,88 +38,115 @@ class ET_EXPERIMENTAL IOManager {
-      executorch::ET_RUNTIME_NAMESPACE::Method& prefill_method,
-      executorch::ET_RUNTIME_NAMESPACE::Method& decode_method) {
-    (void)program;
+      const std::string& prefill_method,
```
Check the prefill_method and decode_method methods exists in module. If there's a typo in the method name, it will be hard to catch it.

shoumikhin replied:
Module will return an error if IOManager tries to work with a Method that doesn't exist, so it's caught at the Module level.

mergennachin commented on extension/llm/runner/io_manager/io_manager.h:
```diff
@@ -38,88 +38,115 @@ class ET_EXPERIMENTAL IOManager {
   ET_NODISCARD virtual runtime::Error reset(
-      executorch::ET_RUNTIME_NAMESPACE::Method& prefill_method,
-      executorch::ET_RUNTIME_NAMESPACE::Method& decode_method) {
+      const std::string& prefill_method,
```
also do validation on the method names

mergennachin commented:
Also fix the unit test in executorch/extension/llm/runner/test:test_text_decoder_runner

shoumikhin committed:
Make IOManager use Module instead of Method. (#13542)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/13542

Let's not expose Method from Module so that it's not getting misused beyond its owner.

Reviewed By: mergennachin

Differential Revision: D80595261
